README.md
# myshift2teams
## Simple API service for organization of shifts

### What?
This is a minimal app sending messages about shifts in a organization to MS Teams. You can send a json to the *myshift* endpoint with a list of people and it will store the configuration and the initial shift in a [Redis](https://redis.io) database. Then you can update the shift or send a shift reminder to a MS Teams channel.

### How to build it?
You will need a D compiler installed (DMD/ldc/...), a C compiler and the DUB package manager. You also will need the sources of phobos library, zlib and openssl lib.

With all these elements, you only need to run `dub install` in this directory.

Alternatively, there is a [Dockerfile](Dockerfile) that you can use to build a container.

### How to run it?
In order to run myshift2teams, you must have installed zlib, openssl library and phobos. And, you'll need a running Redis database where the configuration and shifts will be stored.

Then you can execute the binary:
```
$ ./myshift2teams 
[main(----) INF] Listening for requests on http://127.0.0.1:9000
```
By default, the service will listen at 127.0.0.1:9000 but you can override this value exporting MYSHIFT\_HOST and MYSHIFT\_PORT environment variables. Also you can override the default redis address (127.0.0.1:6379) exporting REDIS\_HOST and REDIS\_PORT environment variables.

You also can use the Docker image, vg.:
```
$ docker pull frantsao/myshift2teams:latest
$ docker run -p 9000:9000 -e MYSHIFT_HOST=0.0.0.0 frantsao/myshift2teams:latest
```

Then you can send a POST to /api/myshift path with a json payload like [data.json](test/data.json). With this configuration example, the hook will be sent to a test endpoint in the same myshift service; you must change the notificationUrl with the webhook url you created in your MS Teams channel:

```
$ http POST 127.0.0.1:9000/api/myshift < data.json 
HTTP/1.1 200 OK
Content-Length: 54
Content-Type: application/json
Date: Tue, 24 May 2022 20:22:12 GMT
Keep-Alive: timeout=10
Server: vibe.d/1.16.0

{
    "OK": "Created/updated configuration"
}
```

The initial config sets the shift to the first name of the list (that is the number 0). You can set that variable to a different number:

```
$ http "127.0.0.1:9000/api/setmyshift?shift=4"
HTTP/1.1 200 OK
Content-Length: 16
Content-Type: application/json; charset=UTF-8
Date: Tue, 24 May 2022 20:24:02 GMT
Keep-Alive: timeout=10
Server: vibe.d/1.22.3

"Set shift to 4"

```

Or you simply need to move along the list:

```
$ http 127.0.0.1:9000/api/updatemyshift
HTTP/1.1 200 OK
Content-Length: 16
Content-Type: application/json; charset=UTF-8
Date: Tue, 24 May 2022 20:26:36 GMT
Keep-Alive: timeout=10
Server: vibe.d/1.22.3

"Set shift to 5"

```

Then you will send the shift reminder:

```
$ http 127.0.0.1:9000/api/sendmyshift
HTTP/1.1 200 OK
Content-Length: 24
Content-Type: application/json; charset=UTF-8
Date: Tue, 24 May 2022 20:29:11 GMT
Keep-Alive: timeout=10
Server: vibe.d/1.22.3

"Sent shift reminder to: Sergio"

```

You always can check the current state of shifts:

```
$ http 127.0.0.1:9000/api/myshift
HTTP/1.1 200 OK
Content-Length: 97
Content-Type: application/json; charset=UTF-8
Date: Mon, 30 May 2022 22:54:06 GMT
Keep-Alive: timeout=10
Server: vibe.d/1.22.3

{
    "Current shift for": "Álex",
    "Team": "[Jaime, Fran, Álex, Marcos, Marcelo, Sergio, Santi, Omar]"
}
```

I sent the example requests  using the wonderful [HTTPie](https://httpie.io/) (but you can use [cURL](https://curl.se/) or whatever).

### Why?

This is a minimal evolution of [my first experiment with Dlang](https://github.com/frantsao/tortilla2teams) that solves a need we had in my team at [idealista](https://idealista.com).

### How?
This has been my documentation:
- [The D language programming web](https://dlang.org/)
- [The vibe.d library web](https://vibed.org/)
- [The Tiny Redis library web](http://adilbaig.github.io/Tiny-Redis/)
- [The Programming in D book](https://ddili.org/ders/d.en/index.html)
- [The Microsoft Teams webhooks documentation](https://docs.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/connectors-using)

### Status
Functional :-)

### Built With

![DMD](https://img.shields.io/badge/LDC-1.29.0-green.svg)
![DUB](https://img.shields.io/badge/DUB-1.28.0-green.svg)
![vibe.d](https://img.shields.io/badge/vibe.d-0.9.4-green.svg)
![tinyredis](https://img.shields.io/badge/tinyredis-2.3.1-green.svg)

### About
![AGPLv3](https://img.shields.io/badge/License-AGPLv3-orange)

This work is under AGPLv3 license (see see the [LICENSE](LICENSE) file).


CHANGELOG.md
# Change Log

## [0.0.1](https://github.com/frantsao/myshift2teams/tree/0.0.1)
### Added
- *First version* @frantsao


source/app.d
import vibe.http.server;
import vibe.http.router;
import vibe.http.client;
import vibe.stream.operations;
import vibe.web.rest;
import vibe.data.json;
import vibe.data.serialization;
import vibe.core.core : runApplication;
import vibe.core.log;
import std.typecons;
import std.array;
import std.algorithm;
import std.string;
import std.conv;
import std.random;
import std.json;
import std.process : environment;
import tinyredis;

@path("/api/") interface MyShiftAPI {
	Json getMyshift();
	Json addMyshift(@viaBody("team") string[] team, 
			@viaBody("notificationUrl") string notificationUrl,
			@viaBody("avatarUrl") string avatarUrl,
			@viaBody("themeColor") string themeColor,
			@viaBody("titleMessage") string titleMessage,
			@viaBody("bodyMessage") string bodyMessage);
	string getSetmyshift(@viaQuery("shift") int shift);
	string getUpdatemyshift();
	Json getSendmyshift();
	@safe string getHealth();
}

class MyShiftAPIService : MyShiftAPI {

	Json getMyshift()
	{
		auto redis_host = environment.get("REDIS_HOST", "127.0.0.1");
    		auto redis_port = to!ushort(environment.get("REDIS_PORT", "6379"));
		auto redis = new Redis(redis_host, redis_port);
		auto currentShift=to!int(redis.send("GET", "shift"));
		auto buddy=to!string(redis.send("LINDEX", "team", currentShift));
		auto team=to!string(redis.send("LRANGE", "team", "0", "-1"));

		return(Json(["Team": Json(team),"Current shift for": Json(buddy)]) );
	}

	string getSetmyshift(@viaQuery("shift") int shift)
	{
 		auto redis_host = environment.get("REDIS_HOST", "127.0.0.1");
    		auto redis_port = to!ushort(environment.get("REDIS_PORT", "6379"));
		auto redis = new Redis(redis_host, redis_port);
		auto maxShift=to!int(redis.send("LLEN", "team"));
		if (( 0 <= shift ) && (shift < maxShift )) {
			redis.send("SET", "shift", shift);
			redis.send("QUIT");
			return(format("Set shift to %d", shift));
		} else {
			redis.send("QUIT");
			return(format("Shift position %d doesn't exist", shift));
		}
	}

	string getUpdatemyshift()
	{
		int shift;
 		auto redis_host = environment.get("REDIS_HOST", "127.0.0.1");
    		auto redis_port = to!ushort(environment.get("REDIS_PORT", "6379"));
		auto redis = new Redis(redis_host, redis_port);
		auto currentShift=to!int(redis.send("GET", "shift"));
		auto maxShift=to!int(redis.send("LLEN", "team"));
		if ( currentShift < maxShift-1 ) {
			shift=currentShift+1;
		} else {
			shift=0;
		}
		redis.send("SET", "shift", shift);
		redis.send("QUIT");
		return(format("Set shift to %d", shift));
	}

	Json addMyshift(@viaBody("team") string[] team,
			@viaBody("notificationUrl") string notificationUrl,
			@viaBody("avatarUrl") string avatarUrl,
			@viaBody("themeColor") string themeColor,
			@viaBody("titleMessage") string titleMessage,
			@viaBody("bodyMessage") string bodyMessage)
	{
		Json returnMessage;

		auto redis_host = environment.get("REDIS_HOST", "127.0.0.1");
   	 	auto redis_port = to!ushort(environment.get("REDIS_PORT", "6379"));
		auto redis = new Redis(redis_host, redis_port);

		redis.send("SET", "shift", "0");
		redis.send("DEL", "team");
		for (int i = 0; i < team.length; ++i) {
			redis.send("RPUSH", "team", team[i]);
		}
		redis.send("SET", "notificationUrl", notificationUrl);
		redis.send("SET", "avatarUrl", avatarUrl);
		redis.send("SET", "themeColor", themeColor);
		redis.send("SET", "titleMessage", titleMessage);
		redis.send("SET", "bodyMessage", bodyMessage);
		redis.send("QUIT");
		returnMessage = Json( "Created/updated configuration" );

		return(returnMessage);
	}
	Json getSendmyshift()
	{
		int shift;
		Json returnMessage;
 		auto redis_host = environment.get("REDIS_HOST", "127.0.0.1");
    		auto redis_port = to!ushort(environment.get("REDIS_PORT", "6379"));
		auto redis = new Redis(redis_host, redis_port);
		auto currentShift=to!int(redis.send("GET", "shift"));
		auto buddy=to!string(redis.send("LINDEX", "team", currentShift));
		auto notificationUrl=to!string(redis.send("GET", "notificationUrl"));
		auto avatarUrl=to!string(redis.send("GET", "avatarUrl"));
		auto themeColor=to!string(redis.send("GET", "themeColor"));
		auto bodyMessage=to!string(redis.send("GET", "bodyMessage"));
		auto titleMessage=to!string(redis.send("GET", "titleMessage"));

		struct Fact {
			string name;
			string value;
		}
		struct Section {
			string activityTitle;
			string activityImage;
			Fact[] facts;
			bool markdown;
		}

		auto fact0 = Fact(bodyMessage, buddy);
		auto section0 = Section(titleMessage, avatarUrl, [fact0], true);
		Json invitationCard = Json.emptyObject;
		invitationCard["@type"] = "MessageCard";
		invitationCard["@context"] = "http://schema.org/extensions";
		invitationCard["summary"] = "Current shift";
		invitationCard["themeColor"]= themeColor;
		invitationCard["sections"]=[section0.serializeToJson()];
		// Would be nice serializing invitationCard from a struct but there is a issue with keys beginning with an '@'
		requestHTTP(notificationUrl,
			(scope req) {
				req.method = HTTPMethod.POST;
				req.writeJsonBody(invitationCard);
			},
			(scope res) {
				logInfo("Webhook URL: %s Status code: %s Response: %s", notificationUrl, res.statusCode, res.bodyReader.readAllUTF8());
			}
		);
		redis.send("QUIT");
		returnMessage=Json(format("Sent shift reminder to: %s",  buddy));
		return(returnMessage);
	}

	@safe string getHealth()
	{
		return("Healthy!");
	}
}

// API for sandboxing webhook endpoints. We can put MS Teams API
//validations here, but we simply send "sections" to the app log

@path("/sandbox/") interface SandboxAPI {
	@safe Json addTest(Json sections);
}


class SandboxAPIService : SandboxAPI {
	@safe Json addTest(Json sections)
	{
		return(sections);
	}
}

// Main process
void main()
{

    auto settings = new HTTPServerSettings;
    auto router = new URLRouter;

    router.registerRestInterface(new MyShiftAPIService);
    router.get("/", (req, res) { res.redirect("/sandbox/test"); } );
    router.get("/api/", (req, res) { res.redirect("/api/myshift"); } );

    router.registerRestInterface(new SandboxAPIService);

    auto host = environment.get("MYSHIFT_HOST", "127.0.0.1");
    auto port = to!ushort(environment.get("MYSHIFT_PORT", "9000"));

    settings.port = port;
    settings.bindAddresses = [host];
    settings.errorPageHandler = (req, res, error)
         {
              with(error) res.writeBody(
              format("Code: %s\n Message: %s\n Exception: %s",
              error.code, 
              error.message, 
              error.exception ? error.exception.msg : "¡Petó como una rata!"));
         };  
    auto l = listenHTTP(settings, router);
    scope (exit) l.stopListening();
    runApplication();
}
